### ISSUE TITLE
Errno::ECONNRESET: Connection reset by peer

### ISSUE THREAD

**Issue by headius** [jruby/jruby#3286]
Transplanted from jruby/jruby-openssl#64 by @green-big-frog.

OpenSSL always fails whit specific versions! Or the handshake misses...

Steps to reproduce:
* Install jruby 1.7.19 or 1.7.22 (other versions I didn't test)
* Do a HTTPS call on an API: `jruby -rjson -r'net/http' -e 'p JSON.parse(Net::HTTP.get(URI.parse("https://api.twitch.tv/kraken/streams/kingkongor")))'`
* You should get following error: https://gist.githubusercontent.com/green-big-frog/df545c1858d410a29625/raw/0e67c659d07f7f7097bd3b4a7ae2acdc8c6415e4/jruby-1.7.22.rb

More details can be found here because I tought it might be an API issue: https://discuss.dev.twitch.tv/t/connection-reset-by-peer-always-the-same-but-just-for-me

**Comment by headius** [comment]
Note from original issue: This works in 9k, which is why we transplanted it from jruby-openssl's issue tracker. I can reproduce it on OS X.

It is also unrelated to json; the error arises from within net/http:

```
[] ~/projects/jruby-1.7 $ jruby -rnet/http -e 'Net::HTTP.get(URI.parse("https://api.twitch.tv/kraken/streams/kingkongor"))'
Errno::ECONNRESET: Connection reset by peer - Connection reset by peer
      read_nonblock at org/jruby/RubyIO.java:2860
          rbuf_fill at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/protocol.rb:141
          readuntil at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/protocol.rb:122
           readline at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/protocol.rb:132
   read_status_line at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:2571
           read_new at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:2560
  transport_request at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:1328
              catch at org/jruby/RubyKernel.java:1242
  transport_request at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:1325
            request at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:1302
        request_get at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:1204
       get_response at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:455
              start at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:746
       get_response at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:454
                get at /Users/headius/projects/jruby-1.7/lib/ruby/1.9/net/http.rb:431
             (root) at -e:1
```

**Comment by headius** [comment]
Ah-ha!

I suppose I should have tested this earlier, but MRI 1.9.3 also exhibits this problem:

```
$ rvm ruby-1.9.3 do ruby -rnet/http -e 'Net::HTTP.get(URI.parse("https://api.twitch.tv/kraken/streams/kingkongor"))'
/Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/protocol.rb:141:in `read_nonblock': Connection reset by peer (Errno::ECONNRESET)
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/protocol.rb:141:in `rbuf_fill'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/protocol.rb:122:in `readuntil'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/protocol.rb:132:in `readline'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:2563:in `read_status_line'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:2552:in `read_new'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:1320:in `block in transport_request'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:1317:in `catch'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:1317:in `transport_request'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:1294:in `request'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:1196:in `request_get'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:455:in `block in get_response'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:746:in `start'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:454:in `get_response'
	from /Users/headius/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/net/http.rb:431:in `get'
	from -e:1:in `<main>'
```

Ruby 2.0 and higher dealt with this by adding buffering and nonblocking read logic to the `openssl` library. Running JRuby with the `--2.0` flag (not recommended in production) allows your script to work properly.

So unfortunately, we can't fix this for you on the 1.7 line, because it would be a behavioral divergence from MRI.

**Comment by green-big-frog** [comment]
I'm totaly fine with using of JRuby 9.0

I just ran in to this issue...

At first it just took me some time to find the origin of the issue.

Thanks for helping me!
